src/backend/auxil/range_alloc.rs

```diff
@@ -118,10 +118,37 @@ where
         self.free_ranges.insert(i, range);
     }
 
+    /// Returns an iterator over allocated non-empty ranges
+    pub fn allocated_ranges<'a>(&'a self) -> impl 'a + Iterator<Item = Range<T>> {
+        let first = match self.free_ranges.first() {
+            Some(Range { ref start, .. }) if *start > self.initial_range.start => Some(self.initial_range.start .. *start),
+            _ => None,
+        };
+
+        let last = match self.free_ranges.last() {
+            Some(Range { end, .. }) if *end < self.initial_range.end => Some(*end .. self.initial_range.end),
+            _ => None,
+        };
+
+        let mid = self.free_ranges
+            .iter()
+            .zip(self.free_ranges.iter().skip(1))
+            .map(|(ra, rb)| ra.end .. rb.start);
```

Comment:
I assume we have the guarantee that these are ordered (appears to be, at least)

Reply:
Yeah, that's what the rest of the code assumes, at least ;)